compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -344,3 +387,65 @@ pub fn create_compressed_metadata_file(
 
     file.write().unwrap()
 }
+
+/// * Xcoff - On AIX, custom sections are merged into predefined sections,
+///   so custom .rustc section is not preserved during linking.
+///   For this reason, we store metadata in predefined .info section, and
+///   define a symbol to reference the metadata. To preserve metadata during
+///   linking on AIX, we have to
+///   1. Create a empty .text section, a empty .data section.
+///   2. Define empty symbol named `symbol_name` inside .data section.
+///   3. Define a symbol named `aix_rust_metadata_symbol_name` referencing
+///      data inside .info section.
+///   From XCOFF's view, (2) creates a csect entry in the symbol table, the
+///   symbol created by (3) is a info symbol for the preceding csect. Thus
+///   two symbols are preserved during linking and we can use the second symbol
+///   to reference the metadata.
```

Comment:
There are two kinds of object files created. One (`create_wrapper_file`) is for use in rlibs and we do not want to preserve metadata in there. It is only wrapped in an rlib to prevent linkers from complaining about files that are not valid object files in archives. The other kind (`create_compressed_metadata_file`) contains compressed metadata and is used for rust dylibs. This one should be preserved.

Reply:
AIX linker will abort if it finds `lib.rmeta` is a valid XCOFF file but only has `.info` section (no `.bss`, `.data` or `.text`). Fortunately, we can just add a stub `.text` section to make linker happy:

```rust
file.add_section(Vec::new(), b".text".to_vec(), SectionKind::Text);
```